src/structures/gdt.rs

```diff
@@ -159,8 +159,11 @@ impl Descriptor {
         // base
         low.set_bits(16..40, ptr.get_bits(0..24));
         low.set_bits(56..64, ptr.get_bits(24..32));
-        // limit (the `-1` in needed since the bound is inclusive)
-        low.set_bits(0..16, (size_of::<TaskStateSegment>() - 1) as u64);
+        // limit (the `-1` is needed since the bound is inclusive)
+        low.set_bits(
+            0..16,
+            (size_of::<TaskStateSegment>() + (tss.iomap_base + iomap_size) as usize - 1) as u64
+        );
```

Comment:
This code only works if the `iomap` is directly adjacent to the `tss`. Also, it makes the function unsafe because setting an `iomap_size` with no iomap would lead to invalid memory accesses.

To solve this, we could create an internal (non-public) `tss_segment_with_limit(tss, limit)` function and move the complete implementation to it. The function would then just do `low.set_bits(0..16, limit - 1)`. On top of that function we can recreate the current `tss_segment` function by calling `tss_segment_with_limit(tss, size_of::<TaskStateSegment>())`.

Now we can create a new `pub unsafe fn tss_segment_with_custom_limit(tss: […], limit: u64)` function that allows taking the `iomap_size` into account. Internally, the function just calls `tss_segment_with_limit(tss, limit)`. This way, the original `tss_segment` function can stay safe and we avoid the breaking parameter addition. This approach also allows us to create a `TssWithIomap` struct in the future and add a safe `tss_segment_with_iomap` function.

What do you think about this?

Reply:
> This code only works if the iomap is directly adjacent to the tss. 

I think that it should work for the TSS anywhere (so long as it is within a 16 bit pointer)? That's why `iomap_base` is added to the size.

> What do you think about this?

Sounds good. Will do.

Reply:
> That's why iomap_base is added to the size.

Sorry, missed that!

> Sounds good. Will do.

Thanks!